### DISCUSSION TITLE
Usage of `mul_add`

### DISCUSSION THREAD

**Discussion by mo8it** [bitshifter/glam-rs#801] [Q&A]
Is there a reason why `mul_add` is not used more often in glam? An example is https://github.com/bitshifter/glam-rs/blob/1ea8163d0907f1ac876d1ec1d213f6add59d9ec0/src/f64/dquat.rs#L659-L662

Using `mul_add` would improve the accuracy and performance (performance on most platforms, see [docs](https://doc.rust-lang.org/stable/std/primitive.f64.html#method.mul_add))

Are you open to PRs that utilize `mul_add`?

**Comment by bitshifter** [comment]
It's unfortunately not as clear a win as it might seem on the surface. From the docs:

> Using mul_add may be more performant than an unfused multiply-add if the target architecture has a dedicated fma CPU instruction. However, this is not always true, and will be heavily dependant on designing algorithms with specific target hardware in mind.

FMA is not enabled by default (on x86_64 at least), it needs to be enabled by users (e.g. with `-C target-feature=+fma`). If FMA is not enabled using `mul_add` will be considerably slower - you can see the different asm generated here https://rust.godbolt.org/z/1WEd9hqTP. If FMA is not available there will be jump to an IEEE-754 conformant implementation that probably looks something like this https://github.com/rust-lang/libm/blob/master/src/math/fmaf.rs.

Possibly glam could add an internal conditional `mul_add` which uses FMA if it is enabled but falls back to an unfused `(a * b) + c` if FMA is not enabled. However this would mean `glam` would give different results depending if it was compiled with or without FMA. This would be a problem for users that are interested in determinism (for example if some supported architectures had FMA and some didn't), so it would be necessary to add a feature to disable glam's conditional FMA in the name of determinism. It's possible, but it gets a little complicated.

**Comment by bitshifter** [comment]
These might be a better option once they're stabilised as they're a hint to the compiler so it can perform more optimizations so there shouldn't be a slow fallback path like there is with `mul_add` https://github.com/rust-lang/rust/issues/136469.
